### ISSUE TITLE
Question: Does Nodebb have a read only mode?

### ISSUE THREAD

**Issue by uplift** [NodeBB/NodeBB#6929]
I know it has a maintenance mode where users are redirected to a holding page but is there a read only mode where users can just read the forum?

Whenever I run certain upgrade scripts, the time it takes to complete sometimes runs into hours due to the size of our database e.g. the session-sharing upgrade script currently takes 3 hours. If I was to run this on our live site it would mean bringing the forum offline for that time. Ideally it would be good to put the forum in read only mode and upgrade a copy of the database, allowing users the ability to access the content then when the upgrade is complete switch to the upgraded database. Currently if I don't put the forum in maintenance mode then any forum data added from the start of the upgrade onwards would be lost.

**Comment by julianlam** [comment]
No, but I don't think it would be particularly difficult. We'd only need to store the "can post" and "can reply" privileges for each category and then revoke them temporarily..

**Comment by uplift** [comment]
I guess it would be pretty simple to build a plugin that hooked into the privileges hooks and if read only then make any write privileges false for all users (except admins). Would just have to make sure all privileges have suitable hooks to plugin to

**Comment by julianlam** [comment]
Oh, I think it's useful enough to be added to core, actually :smile:

**Comment by julianlam** [comment]
With respect to SSO and session-sharing, I'd assume we would not process those during read-only mode.

**Comment by barisusakli** [comment]
Voting, chat, all moderators actions (move, pin, lock etc.) Following users, pretty much anything that writes to db should be disabled no? :trollface:

**Comment by julianlam** [comment]
Hah yes, that is correct, and that's what makes this particular ask about 100x more difficult :worried: 

If we're going to truly add a read-only mode that is good enough to maintain data integrity, that's quite difficult to do, but it might not be _too_ hard to actually limit opaque bits of the UX...

**Comment by pitaj** [comment]
Couple things could be done:

- Create a new layer beneath the dbal that disables write operations when a certain config value is enabled
- Disable sockets and add a new middleware to whitelist certain known to be read-only routes when a certain config value is enabled

**Comment by mySebbe** [comment]
Current Core still has no centralized read-only guard, and writes are not limited to visible content actions: sessions, analytics, jobs, sockets, SSO/ActivityPub, notifications, and plugins also write. Is the required guarantee literally zero database writes, or only “no user-generated/moderation state changes while an upgraded copy is prepared”? The former needs a DBAL barrier plus a very small explicit whitelist; the latter can be a middleware/privilege feature. Naming the guarantee is essential before implementation.
